lib/rubygems.rb

```diff
@@ -1336,6 +1336,17 @@ def default_gem_load_paths
 rescue LoadError
 end
 
+# TruffleRuby >= 24 defines REUSE_AS_BINARY_ON_TRUFFLERUBY in defaults/truffleruby.
+# However, TruffleRuby < 24 defines REUSE_AS_BINARY_ON_TRUFFLERUBY directly in its copy
+# of lib/rubygems/platform.rb, so it is not defined if RubyGems is updated (gem update --system).
+# Instead, we define it here in that case, similar to bundler/lib/bundler/rubygems_ext.rb.
+# We must define it here and not in platform.rb because platform.rb is loaded before defaults/truffleruby.
+class Gem::Platform
+  if RUBY_ENGINE == "truffleruby" && !defined?(REUSE_AS_BINARY_ON_TRUFFLERUBY)
+    REUSE_AS_BINARY_ON_TRUFFLERUBY = %w[libv8 libv8-node sorbet-static].freeze
```

Comment:
Should we move these list to TruffleRuby side? It's hard to maintain this list by RubyGems/Bundler team.

Reply:
If I understand correctly, the list of gems is in truffleruby, this is only a fallback if rubygems gets updated.

Reply:
The list is already on TruffleRuby's side.
But currently it's in the vendored copy of lib/mri/rubygems/platform.rb, which means in RubyGems' CI or when installing a new RubyGems it's gone.
https://github.com/oracle/truffleruby/pull/2819/files is fixing that by moving the definition to lib/truffle/rubygems/defaults/truffleruby.rb.
So this code is there because we need it for TruffleRuby < 24, as the comment says.

It's also the same code as in bundler/lib/bundler/rubygems_ext.rb, which is already in master.

Reply:
Is there any reason why not backport https://github.com/oracle/truffleruby/pull/2819 to TruffleRuby 23?

Reply:
It's too late to backport such changes to 23.1.2, and there won't be more 23.x releases after 23.1.2.

Also because this fallback code is already at https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/rubygems_ext.rb#L297-L299 I don't think it's meaningful to make things more complicated & brittle and avoid it here.